core/src/main/java/org/apache/iceberg/BaseOverwriteFiles.java

```diff
@@ -127,8 +147,29 @@ protected void validate(TableMetadata base) {
       }
     }
 
-    if (conflictDetectionFilter != null && base.currentSnapshot() != null) {
-      validateAddedDataFiles(base, startingSnapshotId, conflictDetectionFilter, caseSensitive);
+
+    if (validateNewDataFiles && base.currentSnapshot() != null) {
+      validateAddedDataFiles(base, startingSnapshotId, conflictDetectionFilter(), caseSensitive);
+    }
+
+    if (validateNewDeleteFiles && base.currentSnapshot() != null) {
+      if (rowFilter() != Expressions.alwaysFalse()) {
+        validateNoNewDeleteFiles(base, startingSnapshotId, conflictDetectionFilter(), caseSensitive);
+      } else if (deletedDataFiles.size() > 0) {
```

Comment:
I think that this should not be an `else if`. Instead, it should be an independent `if`. A caller may delete specific files _and_ set a deletion filter and we need to check both cases.

Reply:
I interpret it like this: call `validateNoNewDeletesForDataFiles` iff we replace only individual files as it more precise than checking if there are any deletes matching the filter. Call `validateNewDeleteFiles` in all other cases (like there are both individual files replaced and the overwrite filter is set). When calling `validateNewDeleteFiles`, we can use `rowFilter()` as the conflict detection iff we don't replace any individual files (i.e. `deletedDataFiles.isEmpty()`).

In my view, `validateNoNewDeleteFiles` is a stronger guarantee than `validateNoNewDeletesForDataFiles`.

Reply:
If so, there is no need to call `validateNoNewDeletesForDataFiles` if the other one is called.

Reply:
When would a user want to make that distinction? I think that it comes down to knowing how the underlying check is implemented, but we don't want users making different calls because they want the underlying system to optimize itself. That seems like it will get users into correctness trouble.

Reply:
These are independent ways to configure the overwrite. There is no guarantee that checking the delete filter will also check the specific data files that were deleted.

I agree with you that when we call `validateNoNewDeletesForDataFiles` we should only pass the actual conflict detection filter and not the row filter. That row filter may not be relevant for the deleted files.

Reply:
Updated.